packages/db/tests/models/test-granule-model.js

```diff
@@ -244,11 +475,314 @@ test('GranulePgModel.upsert() will allow a running status to replace a completed
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'running',
+  });
+
+  const [granuleCumulusId] = await granulePgModel.upsertWithExecutionHistory(
+    knex,
+    granule,
+    executionCumulusId
+  );
+
+  const granuleRecord = await granulePgModel.get(
+    knex,
+    granule
+  );
+
+  t.like(
+    granuleRecord,
+    {
+      ...granule,
+      cumulus_id: granuleCumulusId,
+    }
+  );
+  t.deepEqual(
+    await granuleExecutionHistoryPgModel.search(
+      knex,
+      { granule_cumulus_id: granuleCumulusId }
+    ),
+    [{
+      granule_cumulus_id: Number.parseInt(granuleCumulusId, 10),
+      execution_cumulus_id: executionCumulusId,
+    }]
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() handles multiple executions for a granule', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    executionPgModel,
+    granuleExecutionHistoryPgModel,
+    collectionCumulusId,
+    executionCumulusId,
+  } = t.context;
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'completed',
+  });
+
+  const [granuleCumulusId] = await granulePgModel.upsertWithExecutionHistory(
+    knex,
+    granule,
+    executionCumulusId
+  );
+
+  const [secondExecutionCumulusId] = await executionPgModel.create(
+    knex,
+    fakeExecutionRecordFactory()
+  );
+
+  await granulePgModel.upsertWithExecutionHistory(
+    knex,
+    granule,
+    secondExecutionCumulusId
+  );
+
+  const granuleRecord = await granulePgModel.get(
+    knex,
+    granule
+  );
+
+  t.like(
+    granuleRecord,
+    {
+      ...granule,
+      cumulus_id: granuleCumulusId,
+    }
+  );
+  t.deepEqual(
+    await granuleExecutionHistoryPgModel.search(
+      knex,
+      { granule_cumulus_id: granuleCumulusId }
+    ),
+    [executionCumulusId, secondExecutionCumulusId].map((executionId) => ({
+      granule_cumulus_id: Number.parseInt(granuleCumulusId, 10),
+      execution_cumulus_id: executionId,
+    }))
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() works with transaction', async (t) => {
+  const {
+    knex,
+    granulePgModel,
+    collectionCumulusId,
+    executionCumulusId,
+  } = t.context;
+
+  const granule = fakeGranuleRecordFactory({
+    collection_cumulus_id: collectionCumulusId,
+    status: 'running',
+  });
+
+  const [granuleCumulusId] = await knex.transaction(
+    (trx) =>
+      granulePgModel.upsertWithExecutionHistory(
+        trx,
+        granule,
+        executionCumulusId
+      )
+  );
+
+  const [granuleWithHistory] = await granulePgModel.getWithExecutionHistory(
+    knex,
+    granule
+  );
+
+  t.like(
+    granuleWithHistory,
+    {
+      ...granule,
+      cumulus_id: granuleCumulusId,
+      execution_cumulus_ids: [executionCumulusId],
+    }
+  );
+});
+
+test('GranulePgModel.upsertWithExecutionHistory() does not write anything if execution history upsert fails', async (t) => {
```

Comment:
Again, see comment above re: upsert chain and passing in a knex connection instead.

Reply:
https://github.com/nasa/cumulus/pull/2091/commits/6e52529dfbe7e663f1ecfff0073b130a3510d9d2